step-framework-parent/step-framework-collections/src/main/java/step/core/accessors/AbstractAccessor.java

```diff
@@ -131,12 +143,80 @@ private Equals byId(ObjectId id) {
 
 	@Override
 	public T save(T entity) {
+		if (versionedCollectionDriver != null) {
+			versionedCollectionDriver.save(getVersionForEntity(entity));
+		}
 		return collectionDriver.save(entity);
 	}
 
 	@Override
 	public void save(Iterable<T> entities) {
-		collectionDriver.save((Iterable<T>)entities);
+		if (versionedCollectionDriver != null) {
+			List<VersionableEntity> versionedEntities = new ArrayList<>();
+			entities.forEach(e->versionedEntities.add(getVersionForEntity(e)));
+			versionedCollectionDriver.save(versionedEntities);
+		}
+		collectionDriver.save(entities);
+	}
+
+	private VersionableEntity getVersionForEntity(T entity) {
+		Long current = System.currentTimeMillis();
+		Long currentTsGroup = ((long) (current / newVersionThresholdMs)) * newVersionThresholdMs;
+		//If current version of entity is older than 1 minute create a new one
+		String versionId = entity.getCustomField(VERSION_CUSTOM_FIELD, String.class);
+		VersionableEntity versionableEntity = (versionId != null && ((versionableEntity = getVersionById(versionId)) != null)
+				&& versionableEntity.getUpdateGroupTime() == currentTsGroup) ?
+				versionableEntity : new VersionableEntity();
+
+		versionableEntity.setUpdateTime(current);
+		versionableEntity.setUpdateGroupTime(currentTsGroup);
+		versionableEntity.setEntity(entity);
+		entity.addCustomField(VERSION_CUSTOM_FIELD, versionableEntity.getId().toHexString());

```

Comment:
Potential for side-effect. This method updates the field version but doesn't save the changes

Reply:
Didn't change it for now. The method is private and used to save single and list of entities as bulk. It is implemented this way to still be able to save as bulks, but I will at least add a comment.